### PR TITLE
Add 'item' visual type

### PR DESCRIPTION
**Pull request by GreenXenith**
Fixes https://github.com/minetest/minetest/issues/6986

First pull request for Minetest, and I am a beginner in C++; expect things to be wrong (for example, my commit message is probably too short.)

The `node` visual type for entities mirrors `wielditem` behavior without checking for `wield_image`.

Some test code:
```lua
minetest.register_entity("test:node", {
	visual = "node",
	textures = {"default:fence_wood"},
})

minetest.register_entity("test:wield", {
	visual = "wielditem",
	textures = {"default:fence_wood"},
})
```

### CODE REVIEW

File: src/wieldmesh.h [hunk 1]
```diff
@@ -80,6 +80,7 @@ class WieldMeshSceneNode : public scene::ISceneNode
 	void setCube(const ContentFeatures &f, v3f wield_scale);
 	void setExtruded(const std::string &imagename, const std::string &overlay_image,
 			v3f wield_scale, ITextureSource *tsrc, u8 num_frames);
+	void setItemCheck(const ItemStack &item, Client *client, bool check_wield_image);
 	void setItem(const ItemStack &item, Client *client);
```
Comment (rubenwardy) [inline src/wieldmesh.h]:
```cpp
inline void setItem(const ItemStack &item, Client *client) { setItem(item, client, true); }
```

File: src/wieldmesh.h [hunk 2]
```diff
@@ -80,6 +80,7 @@ class WieldMeshSceneNode : public scene::ISceneNode
 	void setCube(const ContentFeatures &f, v3f wield_scale);
 	void setExtruded(const std::string &imagename, const std::string &overlay_image,
 			v3f wield_scale, ITextureSource *tsrc, u8 num_frames);
+	void setItemCheck(const ItemStack &item, Client *client, bool check_wield_image);
```
Comment (rubenwardy) [inline src/wieldmesh.h]: Rename to `setItem`

File: src/wieldmesh.cpp [hunk 3]
```diff
@@ -445,6 +446,12 @@ void WieldMeshSceneNode::setItem(const ItemStack &item, Client *client)
 	changeToMesh(nullptr);
 }
 
+void WieldMeshSceneNode::setItem(const ItemStack &item, Client *client)
```
Comment (rubenwardy) [inline src/wieldmesh.cpp]: remove

File: doc/lua_api.txt [hunk 4]
```diff
@@ -5486,6 +5486,7 @@ Used by `ObjectRef` methods. Part of an Entity definition.
         --   of its texture.
         --   Otherwise for non-node items, the object will be an extrusion of
         --   'inventory_image'.
+		-- "node" mirrors behavior of "mesh" without the 'wield_image' check.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Please use spaces to indent in the documentation.

Comment (GreenXenith) [reply doc/lua_api.txt]: Whoops. Fixed.
I'll adjust the wording once a solid term is chosen.

**Comment by GreenXenith** [comment]
Leaving as `node` unless someone suggests a better term.

**Comment by stujones11** [comment]
While this is a nice enough feature in itself, to be really useful for falling nodes, etc. it should also support `facedir` as per the original request.

> Ideally, param2 is supported, too.

Now we do actually have a PR for full entity rotation (cough) that would not update the shading. However, we do already have c++ methods to do this properly like `rotateMeshBy6dFacedir` the only question would be how to pass that parameter to the object.

File: src/content_cao.cpp [hunk 5]
```diff
@@ -645,7 +645,13 @@ void GenericCAO::addToScene(ITextureSource *tsrc)
 		}
 		m_wield_meshnode = new WieldMeshSceneNode(
 			RenderingEngine::get_scene_manager(), -1);
-		m_wield_meshnode->setItem(item, m_client);
```
Comment (stujones11) [inline src/content_cao.cpp]:
```c++
		m_wield_meshnode->setItem(item, m_client,
			(m_prop.visual == "wielditem"));
```

File: doc/lua_api.txt [hunk 6]
```diff
@@ -5486,6 +5486,7 @@ Used by `ObjectRef` methods. Part of an Entity definition.
         --   of its texture.
         --   Otherwise for non-node items, the object will be an extrusion of
         --   'inventory_image'.
+	    -- "node" is similar to "wielditem" but ignores the 'wield_image' parameter.
```
[the remark this answers is not in the document]
Comment (GreenXenith) [reply doc/lua_api.txt]: Wait wtf I swear I fixed that

[the remark this answers is not in the document]
Comment (GreenXenith) [reply doc/lua_api.txt]: Its fine on my end, hopefully it will be fixed next commit.

**Comment by GreenXenith** [comment]
So if shading isn't preserved, why bother? Shouldn't we just wait for entity rotation?

And are we good with using `item` instead of `node`?

**Comment by stujones11** [comment]
> Shouldn't we just wait for entity rotation?

Entity rotation will be a very messy way to do this in Lua, unless someone comes up with a clean way to convert `facedir` to a vector of radians ;)

**Comment by GreenXenith** [comment]
> unless someone comes up with a clean way to convert facedir to a vector of radians ;)

See https://github.com/minetest/minetest/pull/7395#issuecomment-429449226.

**Comment by paramat** [comment]
> to be really useful for falling nodes, etc. it should also support facedir as per the original request.

I disagree with adding support for facedir to entity code, just for the sake of the one use case of falling nodes, seems unjustified. Seems better to just use some maths in falling.lua to set the correct object yaw from facedir.
So i think this PR is fine as it is. We can always develop it further later if we decide to.

**Comment by stujones11** [comment]
Okay, `facedir_to_dir` actually returns a unit-vector but I guess you could use that along with some factors of `PI` to get the actual rotation vector.

This feature could always be added later, especially if we find a way to preserve the shading. I was only trying to point out a simple way you could provide all of the requested features and possibly simplify how falling nodes are handled.

**Comment by GreenXenith** [comment]
Im sure the facedir/dir conversions will have to be modified to convert to pitch and roll as well because currently they just do yaw.

**Comment by paramat** [comment]
Yes i mean, we can add a helper function to builtin for facedir to object rotation.

**Comment by stujones11** [comment]
> just for the sake of the one use case of falling nodes, seems unjustified.

It would be useful in any situation where you want to temporarily convert a node into a mesh, the `facedir` property would only be relevant for the wielditem & node/item drawtype, just as `mesh` is only relevant for meshes.

**Comment by GreenXenith** [comment]
> the `facedir` property would only be relevant for the wielditem & node/item drawtype

When the facedir/dir conversions are changed to support pitch/roll and work for any entity, now you have duplicate functionality.

**Comment by stujones11** [comment]
> now you have duplicate functionality.

Yes, but at much greater expense. I agree that you should keep this out of this PR, I may eventually make one for `facedir` support if I can figure out how to preserve the shading.

File: doc/lua_api.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (GreenXenith) [reply doc/lua_api.txt]: Yep, fixed now.

**Comment by paramat** [comment]
> It would be useful in any situation where you want to temporarily convert a node into an object,

Yes, i'm unsure about my opinion now. Edited my comments.

**Comment by GreenXenith** [comment]
Afaik this PR is done, unless I need to fix the 2-characters-too-long line Travis is complaining about.

**Comment by GreenXenith** [comment]
>  if this line goes over 80 columns then yes it's best split

There are like 5 other lines that aren't mine longer than 80, but I changed mine just to make LINT happy.
\*shrug*

**Comment by paramat** [comment]
Yes as we work on MT code we are gradually splitting the long lines for the lines we work on, so it is often inconsistent.

File: src/content_cao.cpp [hunk 7]
```diff
@@ -645,7 +645,15 @@ void GenericCAO::addToScene(ITextureSource *tsrc)
...
+		m_wield_meshnode->setItem(item, m_client,
+			(m_prop.visual == "wielditem"));
+		
+		// Does wield_image need to be checked or not
+		if (m_prop.visual == "wielditem") {
```
Comment (SmallJoker) [inline src/content_cao.cpp]: This code was replaced by L648 & L649. Forgot to delete perhaps?

Comment (GreenXenith) [reply src/content_cao.cpp]: Apparently. That, or I just didn't notice (likely).

**Comment by SmallJoker** [comment]
It's already an improvement. However, I'd also be open for paramtype2 support.

**Comment by GreenXenith** [comment]
Yes, I am starting to see some need for the various `paramtype2`s. `facedir` is probably a must at this point, as well as `wallmounted` and similar. I personally have a need for a way to apply `color`* param2s to entities, so that might be in order as well.

**Comment by paramat** [comment]
See #7897 this visual type should auto-multiply 'visual scale' so that the entity matches a node in scale.

**Comment by GreenXenith** [comment]
> auto-multiply 'visual scale' so that the entity matches a node in scale

Problem being, this only happens for wield_image and nodebox.
I think meshes are a different scale too.

**Comment by paramat** [comment]
Yes, this scaling issue needs discussion.

**Comment by GreenXenith** [comment]
IMO this is an entity problem, considering it's the entities that start out at inconsistent scales. I think the visual types should be consistent from the start, because multiplying it through this feels like a hack. Especially considering the other entity types would need the conversion as well.

**Comment by GreenXenith** [comment]
Once I fix branch conflicts, can this be merged as-is? Things like paramtype2 support should be in a separate PR IMO.

**Comment by GreenXenith** [comment]
Once I fix branch conflicts, can this be merged as-is? Things like paramtype2 support should be in a separate PR IMO.

**Comment by paramat** [comment]
Conflicts can't be viewed as far as i know, you have to compare your PR against recent commits.

**Comment by SmallJoker** [comment]
@GreenXenith Use `git rebase upstream/master` (assuming minetest/minetest is "upstream") and it will tell you whether there are conflicts. GitHub is probably simulating a merge with the target branch whenever you commit and uses the conflict output to display on the webpage. Most likely you won't need to do anything else than running the rebase & force push commands.

Checked it here: `git am -3` applies the patch just fine.

**Comment by GreenXenith** [comment]
I `git fetch`ed and it still works so I guess everything is good.

**Pull request merged**